File: backend/src/services/rule-set.service.ts
import { Injectable } from '@nestjs/common';
import { DataSource, EntityManager } from 'typeorm';
import { now } from '../domain/common/timestamp';
import { TYPE_SAST } from '../domain/issues/types';
import { STATE_OPEN } from '../domain/gate/policy-gate';
import {
    InvalidRuleSetError,
    type StoredRuleFile,
    type TriageImpact,
    type UploadedRuleFile,
    acceptUpload,
    hashRuleSet,
    ruleIdsOf,
    triageImpact
} from '../domain/rules/rule-set';
import { Issue, SemgrepRuleSet } from '../persistence/entities';

/**
 * Uploaded Semgrep rule sets: storing them, activating one, and serving it to executors.
 *
 * **Why this exists rather than the environment variable alone.**
 * `ZANSHIN_SEMGREP_RULES_DIR` is read by the process that scans, and `ScanRunner` is shared
 * between the built-in worker and every remote agent. The directory therefore has to be
 * provisioned on each agent's filesystem, and the control plane has no way to check that it
 * was. Two agents, one provisioned and one not, taking turns on the same target make the
 * SAST backlog resolve and reappear with each turn — silently, because the step *ran* both
 * times. A set stored here and fetched by every executor removes that asymmetry.
 *
 * The environment variable is not withdrawn: it remains the right answer for a
 * single-instance deployment that already manages a volume, and removing it would break
 * those. The precedence is settled in `ScanRunner`, and stated there.
 */
@Injectable()
export class RuleSetService {
    constructor(private readonly dataSource: DataSource) {}

    /**
     * Stores an upload. **Does not activate it.**
     *
     * The two steps are separate because activation is the destructive one: it changes what
     * the next scan looks for, and a rule that disappears takes its open issues with it. An
     * operator uploads, reads what activation would cost, and then decides.
     */
    async store(files: UploadedRuleFile[], name: string, uploadedBy: string | null): Promise<SemgrepRuleSet> {
        const stored = acceptUpload(files);
        const label = (name ?? '').trim();
        if (!label) throw new InvalidRuleSetError('A rule set needs a name.');

        const row = new SemgrepRuleSet();
        row.name = label;
        row.files = stored;
        row.contentHash = hashRuleSet(stored);
        row.ruleCount = ruleIdsOf(stored).size;
        row.fileCount = stored.length;
        row.sizeBytes = String(stored.reduce((total, file) => total + Buffer.byteLength(file.content, 'utf8'), 0));
        // `null`, not `false`: the unique index over this column is what enforces "at most
        // one active", and it only counts NULLs as distinct.
        row.isActive = null;
        row.uploadedBy = uploadedBy;
        row.uploadedAt = now();
        row.activationNote = null;

        return this.dataSource.manager.save(row);
    }

    /** Every stored set, newest first, without their content. */
    async list(): Promise<Omit<SemgrepRuleSet, 'files'>[]> {
        const rows = await this.dataSource.getRepository(SemgrepRuleSet).find({ order: { uploadedAt: 'DESC' } });
        // The files are megabytes; a listing that carried them would be a listing nobody
        // opens twice.
        return rows.map(({ files: _files, ...rest }) => rest as Omit<SemgrepRuleSet, 'files'>);
    }

    /** The active set, or `null` when only the bundled rules apply. */
    async active(): Promise<SemgrepRuleSet | null> {
        return this.dataSource.getRepository(SemgrepRuleSet).findOne({ where: { isActive: true } });
    }

    /** One stored set, files included. */
    async byId(id: number): Promise<SemgrepRuleSet | null> {
        return this.dataSource.getRepository(SemgrepRuleSet).findOneBy({ id });
    }

    /** A set by the hash an executor holds, for the fetch route. */
    async byHash(contentHash: string): Promise<SemgrepRuleSet | null> {
        return this.dataSource.getRepository(SemgrepRuleSet).findOne({ where: { contentHash } });
    }

    /**
     * What activating this set would do to the existing backlog.
     *
     * **The answer an operator has to see before clicking.** A rule id enters an issue's
     * fingerprint: a rule that is not in the new set stops being found, its issues are
     * resolved on the next scan, and their triage decisions — justifications, review dates,
     * who decided — go with them. Nothing errors, and the dashboard looks better afterwards.
     *
     * Counted from the open SAST issues that exist right now, grouped by the rule that
     * produced them, because the backlog is the authority on what has something to lose —
     * not the previously uploaded set, since rules also arrive from the bundled tree and
     * from `ZANSHIN_SEMGREP_RULES_DIR`.
     */
    async impactOf(candidate: SemgrepRuleSet): Promise<TriageImpact> {
        const current = await this.active();
        const currentIds = current ? ruleIdsOf(current.files as StoredRuleFile[]) : new Set<string>();
        const nextIds = ruleIdsOf(candidate.files as StoredRuleFile[]);

        return triageImpact(currentIds, nextIds, await this.openSastIssuesByRule());
    }

    /**
     * Activates a set, and records what the operator was told it would cost.
     *
     * In one transaction with the deactivation: the unique index makes two active rows
     * impossible, so a partial application would leave *none* active — silently falling back
     * to the bundled rule, which is the failure this whole feature exists to prevent.
     */
    async activate(id: number, note: string | null): Promise<SemgrepRuleSet> {
        return this.dataSource.transaction(async (manager: EntityManager) => {
            const target = await manager.findOneBy(SemgrepRuleSet, { id });
            if (!target) throw new InvalidRuleSetError(`No rule set with id ${id}.`);

            await manager.update(SemgrepRuleSet, { isActive: true }, { isActive: null });
            await manager.update(SemgrepRuleSet, { id }, { isActive: true, activationNote: note });

            return (await manager.findOneBy(SemgrepRuleSet, { id })) as SemgrepRuleSet;
        });
    }

    /** Returns to the bundled rules alone. */
    async deactivateAll(): Promise<void> {
        await this.dataSource.manager.update(SemgrepRuleSet, { isActive: true }, { isActive: null });
    }

    /**
     * Open SAST issues, counted per rule identifier.
     *
     * Read from `t_issue` alone: the issue carries its own `type` and `identifier`, and that
     * identifier **is** Semgrep's `check_id` — the same string, because
     * `--no-rewrite-rule-ids` stops Semgrep prefixing it with the rule file's path. Joining
     * `t_finding` to reach it would add a join for a column that is already here, and would
     * count an issue once per scan that saw it.
     */
    private async openSastIssuesByRule(): Promise<Map<string, number>> {
        const rows: { identifier: string | null; count: string }[] = await this.dataSource
            .getRepository(Issue)
            .createQueryBuilder('issue')
            .select('issue.identifier', 'identifier')
            .addSelect('COUNT(issue.id)', 'count')
            .where('issue.state = :state', { state: STATE_OPEN })
            .andWhere('issue.type = :type', { type: TYPE_SAST })
            .groupBy('issue.identifier')
            .getRawMany();

        const counts = new Map<string, number>();
        for (const row of rows) {
            // `getRawMany` short-circuits hydration: the count arrives as the driver returns
            // it, which is a string on PostgreSQL and MySQL. Reading it as a number without
            // converting would compare `"12"` against numbers elsewhere.
            if (row.identifier) counts.set(row.identifier, Number(row.count));
        }
        return counts;
    }
}

File: zanshin-java/zanshin-core/src/main/java/com/asmolabs/zanshin/core/services/RuleSetService.java
package com.asmolabs.zanshin.core.services;

import com.asmolabs.zanshin.common.domain.issues.FindingType;
import com.asmolabs.zanshin.common.domain.issues.IssueState;
import com.asmolabs.zanshin.common.domain.rules.InvalidRuleSetException;
import com.asmolabs.zanshin.common.domain.rules.RuleSet.StoredFile;
import com.asmolabs.zanshin.common.domain.rules.RuleSet.TriageImpact;
import com.asmolabs.zanshin.common.domain.rules.RuleSet.UploadedFile;
import com.asmolabs.zanshin.common.domain.rules.RuleSet;
import com.asmolabs.zanshin.core.persistence.SemgrepRuleSetEntity;
import com.asmolabs.zanshin.core.repositories.Issues;
import com.asmolabs.zanshin.core.repositories.RuleSetSummary;
import com.asmolabs.zanshin.core.repositories.RuleSets;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.charset.StandardCharsets;
import java.time.Clock;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Uploaded Semgrep rule sets: storing them, activating one, and serving them to executors.
 *
 * <p><b>Why this exists rather than the environment variable alone.</b> {@code
 * ZANSHIN_SEMGREP_RULES_DIR} is read by the process that scans, and the scanner is shared
 * between the built-in worker and every remote agent. The directory therefore has to be
 * provisioned on each agent's filesystem, and the control plane has no way to check that it
 * was. Two agents, one provisioned and one not, taking turns on the same target make the SAST
 * backlog resolve and reappear with each turn — silently, because the step <em>ran</em> both
 * times. A set stored here and fetched by every executor removes that asymmetry.
 *
 * <p>The environment variable is not withdrawn: it stays the right answer for a
 * single-instance deployment that already manages a volume. The precedence between the two is
 * settled in the scanner, and stated there.
 */
@Service
public class RuleSetService {

    private static final TypeReference<List<StoredFile>> FILES = new TypeReference<>() {};

    private final RuleSets ruleSets;
    private final Issues issues;
    private final ObjectMapper json;
    private final Clock clock;

    public RuleSetService(RuleSets ruleSets, Issues issues, ObjectMapper json, Clock clock) {
        this.ruleSets = ruleSets;
        this.issues = issues;
        this.json = json;
        this.clock = clock;
    }

    /**
     * Stores an upload. <b>Does not activate it.</b>
     *
     * <p>The two steps are separate because activation is the destructive one: it changes what
     * the next scan looks for, and a rule that disappears takes its open issues with it. An
     * operator uploads, reads what activation would cost, and then decides.
     */
    @Transactional
    public SemgrepRuleSetEntity store(List<UploadedFile> files, String name, String uploadedBy) {
        String label = name == null ? "" : name.trim();
        if (label.isEmpty()) {
            throw new InvalidRuleSetException("A rule set needs a name.");
        }

        List<StoredFile> stored = RuleSet.accept(files);

        SemgrepRuleSetEntity row = new SemgrepRuleSetEntity();
        row.setName(label);
        row.setFiles(writeFiles(stored));
        row.setContentHash(RuleSet.contentHash(stored));
        row.setRuleCount(RuleSet.ruleIdsOf(stored).size());
        row.setFileCount(stored.size());
        row.setSizeBytes(stored.stream()
                .mapToLong(file -> file.content().getBytes(StandardCharsets.UTF_8).length)
                .sum());
        // `null`, not `false`. See `RuleSets`: the unique index is the guard.
        row.setIsActive(null);
        row.setUploadedBy(uploadedBy);
        row.setUploadedAt(clock.instant());
        row.setActivationNote(null);

        return ruleSets.save(row);
    }

    /** Every stored set, newest first, without their content. */
    @Transactional(readOnly = true)
    public List<RuleSetSummary> list() {
        return ruleSets.summaries();
    }

    /** The active set, or empty when only the bundled rules apply. */
    @Transactional(readOnly = true)
    public Optional<SemgrepRuleSetEntity> active() {
        return ruleSets.findByIsActiveTrue();
    }

    @Transactional(readOnly = true)
    public Optional<SemgrepRuleSetEntity> byId(long id) {
        return ruleSets.findById(id);
    }

    /** A set by the hash an executor holds, for the fetch route. */
    @Transactional(readOnly = true)
    public Optional<SemgrepRuleSetEntity> byHash(String contentHash) {
        return ruleSets.findFirstByContentHashOrderByIdAsc(contentHash);
    }

    /** The rules of a stored set, decoded. */
    public List<StoredFile> filesOf(SemgrepRuleSetEntity row) {
        try {
            return json.readValue(row.getFiles(), FILES);
        } catch (JsonProcessingException corrupt) {
            // Distinct from "no set is active": falling back to the bundled rule here would
            // silently narrow what every scan looks for, which is the failure this whole
            // feature exists to prevent.
            throw new InvalidRuleSetException("The stored rules of set " + row.getId() + " cannot be read back.");
        }
    }

    /**
     * What activating this set would do to the existing backlog.
     *
     * <p><b>The answer an operator has to see before clicking.</b> A rule id enters an issue's
     * fingerprint: a rule absent from the new set stops being found, its issues resolve on the
     * next scan, and their triage — justifications, review dates, who decided — goes with them.
     * Nothing errors, and the dashboard looks better afterwards.
     *
     * <p>Counted from the open SAST issues that exist right now rather than from the previously
     * uploaded set, because the backlog is the authority on what has something to lose: rules
     * also arrive from the bundled tree and from {@code ZANSHIN_SEMGREP_RULES_DIR}.
     */
    @Transactional(readOnly = true)
    public TriageImpact impactOf(SemgrepRuleSetEntity candidate) {
        Set<String> current = active().map(row -> RuleSet.ruleIdsOf(filesOf(row))).orElseGet(Set::of);
        return RuleSet.impact(current, RuleSet.ruleIdsOf(filesOf(candidate)), openSastIssuesByRule());
    }

    /**
     * Activates a set, and records what the operator was told it would cost.
     *
     * <p>In one transaction with the deactivation: the unique index makes two active rows
     * impossible, so a half-applied change would leave <em>none</em> active — silently falling
     * back to the bundled rule, which is precisely the outcome this feature exists to prevent.
     */
    @Transactional
    public SemgrepRuleSetEntity activate(long id, String note) {
        SemgrepRuleSetEntity target = ruleSets
                .findById(id)
                .orElseThrow(() -> new InvalidRuleSetException("No rule set with id " + id + "."));

        ruleSets.deactivateAll();
        ruleSets.activate(target.getId(), note);

        // Re-read rather than mutating the object in hand: the two statements above bypass the
        // persistence context, so the entity loaded before them still says what it said.
        return ruleSets.findById(id).orElseThrow(() -> new InvalidRuleSetException("No rule set with id " + id + "."));
    }

    /** Returns to the bundled rules alone. */
    @Transactional
    public void deactivateAll() {
        ruleSets.deactivateAll();
    }

    /**
     * Open SAST issues, counted per rule identifier.
     *
     * <p>Read from the issues alone: an issue carries its own type and identifier, and that
     * identifier <b>is</b> Semgrep's {@code check_id} — the same string, because {@code
     * --no-rewrite-rule-ids} stops Semgrep prefixing it with the rule file's path.
     */
    private Map<String, Long> openSastIssuesByRule() {
        Map<String, Long> counts = new HashMap<>();
        for (Object[] row : issues.countOpenByIdentifier(IssueState.OPEN.wireName(), FindingType.SAST.wireName())) {
            counts.put((String) row[0], ((Number) row[1]).longValue());
        }
        return counts;
    }

    private String writeFiles(List<StoredFile> files) {
        try {
            return json.writeValueAsString(files);
        } catch (JsonProcessingException impossible) {
            throw new IllegalStateException("A validated rule set could not be serialized", impossible);
        }
    }
}
